.github/workflows/haskell-ci.yml

```diff
@@ -102,8 +102,11 @@ jobs:
       - name: cache (tools)
         uses: actions/cache@v2
         with:
-          key: ${{ runner.os }}-${{ matrix.ghc }}-tools
-          path: ~/.cabal/store-tools
+          key: |
+            ${{ runner.os }}-${{ matrix.ghc }}-tools
+          path: |
+            ~/.cabal/store-tools
+            ~/.hlint
```

Comment:
I'd download to `~/.cabal/store-tools`. That was an idea. The name is bad though. Rather it could be `~/.haskell-ci-tools`

The  `doctest` installation should probably be changed to use `~/.haskell-ci-tools/store` as store, and hlint could be in `~/.haskell-ci-tools/hlint`.

Reply:
I'd be fine with that.

Tangentially related question: when we `cabal v2-install` `doctest` and `hlint`, why do we specify a separate `--store-dir` in the first place? In particular, the Travis and Bash backends don't do this, so I'm wondering why the GitHub Actions backend has this additional complexity.

Reply:
bash script has a luxury of using single store for everything. Doesn't need to worry about uploads/downloads.

If you look closely, the GitHub actions uses two caches,
1. This one for tools which has strict key.
2. Another for actual project

Important bit to know is that GitHub actions **doesn't** re-upload  cache if it's downloaded with exact hit of the key. 
(n fact, we should add something identifying hlint/doctest versions there for tools-cache).

The actual dependencies store cache is separate, and is (unfortunately) completely re-uploaded for every job. We use `restore-key` to get *some* cache, there is prefix-matching logic. There are options, currently we use `${{ github.sha }}` which is globally available. Another alternative is to generate `cabal.project.freeze` and use its hash as cache-key. The latter would be better (then in PRs caches won't be reuploaded, saves some time and bandwidth). But we cannot caclculate cabal.project.freeze before checkout, and tools are installed before that. So that is why two separate caches.

Reply:
Also tools cache is indeed very static, so it makes sense to have it separate. (If not `doctest` dependendcy on GHC, it could be be shared across all jobs!)

Reply:
Good to know, thanks.

I've pushed an updated version that uses `~/.haskell-ci-tools`. Can you double-check to make sure I didn't do something silly in the process of changing the store location?